Update cc_toolchain configuration to be compatible with bazel 7.

zygoloid opened the pull request.

zygoloid committed:
Update cc_toolchain configuration to be compatible with bazel 7.

zygoloid committed:
Remove crosstool_top. It doesn't do anything any more.

zygoloid committed:
Fix up target-based config.

jonmeow commented on .bazelrc:
```diff
@@ -37,6 +30,10 @@ build --experimental_guard_against_concurrent_changes
 # https://github.com/bazelbuild/bazel/issues/13315
 build --incompatible_dont_enable_host_nonhost_crosstool_features=false
 
+# Opt into the new Bazel toolchain resolution mechanism. This is enabled by
+# default in Bazel 7.
```
Maybe an explicit TODO to remove once bazel 7 can be treated as minimum?

jonmeow commented on bazel/cc_toolchains/clang_cc_toolchain_config.bzl:
```diff
@@ -1017,38 +1014,39 @@ def _impl(ctx):
         features.append(default_link_libraries_feature)
     features.append(final_flags_feature)
 
+    identifier = "local-" + "-" + ctx.attr.target_cpu + "-" + ctx.attr.target_os
```
```suggestion
    identifier = "local-{0}-{1}".format(ctx.attr.target_cpu, ctx.attr.target_os)
```

Why `local--`? Is that a typo? Suggestion to remove the issue, make format clearer.

jonmeow commented on bazel/cc_toolchains/clang_cc_toolchain_config.bzl:
```diff
@@ -1101,13 +1100,15 @@ def cc_local_toolchain_suite(name, cpus):
     # Create the individual local toolchains for each CPU.
-    for cpu in cpus:
+    for (os, cpu) in configs:
+        config_name = name + "_" + os + "_" + cpu
```
```suggestion
        config_name = "{0}_{1}_{2}".format(name, os, cpu)
```

Suggesting because there are so many values, mainly due to the above `local--` comment.

zygoloid committed:
Debugging mac build failure

jonmeow commented on bazel/cc_toolchains/clang_cc_toolchain_config.bzl:
```diff
@@ -1117,12 +1118,13 @@ def cc_local_toolchain_suite(name, cpus):
+            toolchain_config = ":" + config_name + "_config",
+            toolchain_identifier = config_name,
+        )
+        native.toolchain(
+            name = config_name,
+            exec_compatible_with = ["@platforms//cpu:" + cpu, "@platforms//os:" + os],
+            target_compatible_with = ["@platforms//cpu:" + cpu, "@platforms//os:" + os],
```
Maybe put the labels in a local to reduce repetition?

```suggestion
        compatible_with = ["@platforms//cpu:" + cpu, "@platforms//os:" + os]
        native.toolchain(
            name = config_name,
            exec_compatible_with = compatible_with,
            target_compatible_with = compatible_with,
```

jonmeow commented on bazel/cc_toolchains/clang_toolchain.BUILD:
```diff
@@ -9,12 +9,12 @@ load(":cc_toolchain_config.bzl", "cc_local_toolchain_suite")
-        "k8",
-        "x64_windows",
+    configs = [
+        ("linux", "aarch64"),
+        ("linux", "x86_64"),
+        ("freebsd", "x86_64"),
+        ("macos", "arm64"),
+        ("macos", "x86_64"),
+        ("windows", "x86_64"),
```
This list repeats in the WORKSPACE. Maybe move it to clang_configuration.bzl for reuse?

jonmeow reviewed (approved):
Approving with some notes, but feel free to merge if you'd rather just get the fix in.

jonmeow commented:
Looks like macos config needs to be examined.

zygoloid committed:
Don't use --start-lib ... --end-lib on Mac.

They don't exist.

zygoloid committed:
Remove debug.

zygoloid committed:
Apply suggestions from code review

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Factor out duplicated list of targets.

jonmeow commented on bazel/cc_toolchains/clang_configuration.bzl:
```diff
@@ -250,3 +260,9 @@ configure_clang_toolchain = repository_rule(
+    configure_clang_toolchain(name = name)
+
+    for os, cpu in clang_configs:
+        native.register_toolchains("@{0}//:bazel_cc_toolchain_{1}_{2}".format(name, os, cpu))
```
```suggestion
        native.register_toolchains(
            "@{0}//:bazel_cc_toolchain_{1}_{2}".format(name, os, cpu)
        )
```

Wrapping nit (I think if you apply this, buildifier will maintain it)

jonmeow reviewed (commented):
Still LGTM

zygoloid committed:
Update bazel/cc_toolchains/clang_configuration.bzl

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Comment update

zygoloid committed:
buildifier

zygoloid committed:
Get mac build to work.

Put back --start-lib / --end-lib. I don't think they're necessary, but
this keeps this change smaller.
